### DISCUSSION TITLE
Recommended typesafe cache update approach

### DISCUSSION THREAD

**Discussion by robertherber** [urql-graphql/urql#2711] [Q&A]
I love the URQL approach of doing optimistic/cache updates in one place. But I still can't quite figure out how to get it quite right, with full type support.

The closest I've come so far:
1. Read data from the cache with `cache.readFragment`. This way I get the typed object (the good), but it's based on a specific fragment which might not have all data (the bad).
2. Do whatever I need to update to reflect the optimistic update.
3. Return this, sometimes typecasting the modified fragment this to the expected full data type (not optimal of course)

What I'd like to do:
1. Read typed data from the cache with a `__typename` and `id`, similarly to `cache.readFragment` but that always returns all available data.
2. Do whatever I need to update to reflect the optimistic update.
3. Return the modified object that contains all available data (not just a fragment). Fully typed and including all available data.

Is there a recommended approach for this?

**Comment by thdxr** [comment]
I need this as well and I'm not sure if this is because I'm misunderstanding urql to some degree. It is a bit annoying to have to return the whole object when I'm just trying to optimistically update one field in it

**Comment by taiki-fr** [comment]
+1 from here as well. I've found a couple comments from @kitten that seem to say that it's actually okay for the `optimistic` function to return a partial object containing only the fields that should be updated optimistically:
- https://github.com/urql-graphql/urql/discussions/1110#discussioncomment-119197
- https://github.com/urql-graphql/urql/discussions/2427#discussioncomment-2666639

This is great news but it also contradicts the docs (https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates), which state that "it's important to ensure that our optimistic mutations return all data that the real mutation may return". As multiple people including the OP have noted, that sentence seems to suggest that our optimistic return will need to list every field that exists in the GraphQL type being updated, in order for the updater to accommodate the mutation being called with different selection sets. Would love some clarification on this!

**Reply by kitten** [reply]
@taiki-fr The docs are a little behind on this. Apologies!

Yes, this limitation has been lifted in part. While it's now possible to return _less_ data in optimistic updates, as needed, it's still important however to do return the data that we want to "override" so to speak.
This was more in line with expected behaviours and implemented here:
- https://github.com/urql-graphql/urql/pull/2616
- with a fix later being applied here: https://github.com/urql-graphql/urql/pull/2740

In short, the docs may have to have updated phrasing about this, but you can indeed return partial optimistic results now

There's also a GCG Typegen module that generates the types needed for (some/most; not all) configs in Graphcache and should help with this (but I don't think it has been updated for this fully yet; we're not advertising it, because we're currently not 100% happy with how much tooling has to be set up to get such a "basic" feature 😅)

**Reply by taiki-fr** [reply]
That's great to hear! Being able to return partial results as well as centralizing the update setup per mutation type is a big win over Apollo's API imo. Thanks for the clarification, keep up the great work!

**Reply by taiki-fr** [reply]
@kitten Follow up question: when you say the limitation has been lifted _in part_, does that just mean you also need to include the `id` and `__typename`, or whatever is needed to determine the key for the entity, so that urql knows which entity to update?
